metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+                logger.warning(message)
+                raise ConnectionError(message)
+
+            if res.json()["status"].upper() == "SUCCEEDED":
+                logger.info(f"Scan result is available for scan id({scan_id})")
+                return True
+
+            if retry == max_retry:
+                logger.warning("Max retry reached. Scan job result is not available")
+                break
+
+            logger.info(f"Sleeping for {minimum_sleep} seconds")
+            sleep(minimum_sleep)
+            retry += 1
+
+        # Result is not available
+        return False
+
+    def get_users(self, workspace_id: str, entity: str, entity_id: str) -> List[User]:
+        """
+        Get user for the given PowerBi entity
+        """
+
+        user_list_endpoint: str = AdminAPIResolver.API_ENDPOINTS[
+            Constant.ENTITY_USER_LIST
+        ]
+        # Replace place holders
+        user_list_endpoint = user_list_endpoint.format(
+            POWERBI_ADMIN_BASE_URL=DataResolverBase.ADMIN_BASE_URL,
+            ENTITY=entity,
+            ENTITY_ID=entity_id,
+        )
+        # Hit PowerBi
+        logger.info(f"Request to URL={user_list_endpoint}")
+        response = self._request_session.get(
+            user_list_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        response.raise_for_status()
+
+        users_dict: List[Any] = response.json()[Constant.VALUE]
+
+        # Iterate through response and create a list of PowerBiAPI.Dashboard
+        users: List[User] = [
+            User(
+                id=instance.get("identifier"),
+                displayName=instance.get("displayName"),
+                emailAddress=instance.get("emailAddress"),
+                graphId=instance.get("graphId"),
+                principalType=instance.get("principalType"),
+            )
+            for instance in users_dict
+        ]
+
+        return users
+
+    def get_scan_result(self, scan_id: str) -> dict:
+        logger.info("Fetching scan result")
+        logger.info(f"{Constant.SCAN_ID}={scan_id}")
+        scan_result_get_endpoint = AdminAPIResolver.API_ENDPOINTS[
+            Constant.SCAN_RESULT_GET
+        ]
+        scan_result_get_endpoint = scan_result_get_endpoint.format(
+            POWERBI_ADMIN_BASE_URL=DataResolverBase.ADMIN_BASE_URL, SCAN_ID=scan_id
+        )
+
+        logger.debug(f"Hitting URL={scan_result_get_endpoint}")
+        res = self._request_session.get(
+            scan_result_get_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+        if res.status_code != 200:
+            message = f"API({scan_result_get_endpoint}) return error code {res.status_code} for scan id({scan_id})"
+            logger.warning(message)
+            raise ConnectionError(message)
+
+        return res.json()["workspaces"][0]
```

Comment:
What if a) workspaces does not exist or b) workspaces is not an array or c) workspaces is empty?

Reply:
This function gets called only if `wait_for_scan_to_complete` return true and hence get_scan_result will always return the data.
Still I have added a guard condition